### PR TITLE
Support `__cuda_array_interface__` in `cupy.poly1d`

### PR DESCRIPTION
**Pull request by Dahlia-Chehata**
Some changes in tests may be needed after `cupy.polymul` is merged

### CODE REVIEW

**Comment by asi1024** [comment]
I don't want implicit data conversion that requires device synchronization. I think it's OK to leave `ndarray + poly1d` unsupported.

**Comment by asi1024** [comment]
I thought a bit more and came to think we can later reduce device synchronization in `poly1d.coeffs`, so we will merge this PR. Could you resolve the conflict?

File: cupy/lib/polynomial.pyx [hunk 1]
```diff
@@ -105,6 +105,23 @@ cdef class poly1d:
...
+            'descr': arr.dtype.descr,
+            'version': 2,
+        }
+        desc['strides'] = None
+        if arr.size > 0:
+            desc['data'] = (arr.data.ptr, False)
+        else:
+            desc['data'] = (0, False)
+
+        return desc
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:123]:
```suggestion
        return arr.__cuda_array_interface__
```

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 2]
```diff
@@ -238,18 +238,34 @@ def _get_input(self, xp, in_type, dtype):
         lambda x, y: x + y,
         lambda x, y: x - y,
     ],
-    'type_l': ['poly1d', 'python_scalar'],
+    'type_l': ['poly1d', 'python_scalar', 'ndarray'],
     'type_r': ['poly1d', 'ndarray', 'python_scalar', 'numpy_scalar'],
 }))
-class TestPoly1dArithmetic(Poly1dTestBase):
+class TestPoly1dArithmetic1(Poly1dTestBase):
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:244]:
```suggestion
class TestPoly1dPolynomialArithmetic(Poly1dTestBase):
```

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 3]
```diff
@@ -238,18 +238,34 @@ def _get_input(self, xp, in_type, dtype):
...
+        return self.func(a1, a2)
+
+
+@testing.gpu
+@testing.parameterize(*testing.product({
+    'fname': ['add', 'subtract', 'multiply', 'divide', 'power'],
+    'type_l': ['poly1d', 'ndarray', 'python_scalar', 'numpy_scalar'],
+    'type_r': ['poly1d'],
+}))
+class TestPoly1dArithmetic2(Poly1dTestBase):
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:260]:
```suggestion
class TestPoly1dMathArithmetic(Poly1dTestBase):
```

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Comment by leofang** [comment]
Nice!

**Pull request merged**